### ISSUE TITLE
Allow reductions to write into pinned memory

### ISSUE THREAD

**Issue by esoha-nvidia** [pytorch/pytorch#109052]
### 🚀 The feature, motivation and pitch

Reduction operations that have small output, like `tensor.max` and `tensor.min`, ought to allow the output to be in pinned host memory if the input is in GPU.

This improves performance of `x.max().item()` by 5-7x.

### Alternatives

One alternative is to just live with it.  The current way is to just let the output be on the GPU and then copy it from GPU to CPU.  That uses a memcpy, even for a tiny tensor of 4 bytes, and it's wildly slow.

Another alternative is to write your own pybind function that is basically just like the pytorch one except without the error message.

Another option is to, instead of using `copy_` to bring the data from the GPU to CPU, make a function to replace `copy_` that can notice that the data to copy is so tiny that it would be faster to launch and kernel and do the copy to pinned memory than it would be to do a memcpyasync.

### Additional context

```python
a = torch.randn(1,3, device="cuda")
output = torch.empty([1], dtype=a.dtype, pin_memory=True)
torch.max(a, out=output)
```

The error is:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected self.device() == out.device() to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

Here is the profiler output:

![image](https://github.com/pytorch/pytorch/assets/69258779/e20daeea-e418-4134-876d-2ced8d8d3ade)

We see that it's spending *most* of the time just copying 4 bytes from GPU to CPU.  In my testing using cub directly, you don't have to wait for the copy, you just need the synchronization and it's super fast, like 5-7x.

You might even consider making pinned memory the default for reductions!  It's really fast.

**Comment by esoha-nvidia** [comment]
https://nvidia.slack.com/archives/C8HLVDB7X/p1694449586341099

**Comment by malfet** [comment]
@ptrblck let's chat about this one during the next syncup. But looks like `module: cuda` and `module: performance` are the right flags here.

**Comment by ptrblck** [comment]
> This improves performance of x.max().item() by 5-7x.
> That uses a memcpy, even for a tiny tensor of 4 bytes, and it's wildly slow.

I don't believe that's the case and think you are just skipping the needed synchronization while `item()` will accumulate the runtime of queued and launched kernels. 
Based on your experiments it seems you are seeing an improvement in performance when an async memcopy is used, which needs a synchronization to avoid race conditions.
I've recently explained the issue and posted a small code snippet showing this race [in this post](https://discuss.pytorch.org/t/memory-pollution-while-concurrent-data-transfer-with-multi-cuda-stream/187801/2?u=ptrblck) which you can use for experimentation.
Let me know if I misunderstand the issue.

**Comment by esoha-nvidia** [comment]
Yes, I think that you might misunderstand the issue. I did the synchronization correctly. The issue was the profiler, which is making memcpy appear slower than writing to pinned memory. Without the profiler, the memcpy is faster. You will not be able to measure this in python, you must write cuda because python cannot reduce to a pinned memory. I measured this with Matthias (mjoux).

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ptrblck** [comment]
> Yes, I think that you might misunderstand the issue. I did the synchronization correctly. The issue was the profiler, which is making memcpy appear slower than writing to pinned memory. Without the profiler, the memcpy is faster.

That's interesting as it also seems you are claiming now copies to pageable memory are faster than to pinned memory?

@malfet Sure, we can discuss it, but I'm utterly confused about claims seeing a 5x speedup and afterwards a slowdown blaming the profiler, so I would need more information what we are trying to achieve.

**Comment by esoha-nvidia** [comment]
With the profiler running, I saw that memcpy to get the result of a reduction was very slow.

![image](https://github.com/pytorch/pytorch/assets/69258779/316843d3-1c4c-4b95-b1d5-de578bd3d8b3)

The kernel is the blue bit that doing the reduction.  After that there is a very long pause, much longer than the reduction, where 4 bytes are copied from GPU memory to pinned memory on the GPU.

If the goal is to get the result of the reduction on the CPU, another way to do it is to give a tensor that is in pinned memory to the reduction, as output.  In my measurements while using the profiler, this was 5-7x times faster.  That really big pause in the image above mostly went away.  So I was under the impression that writing directly to pinned memory was faster.

Then I worked with Matthias and we did some more measurements and it turns out that I was correct but only when the profiler is running.  When the profiler isn't running, the memcpy technique that pytorch is already using is a faster way to get the result of the reduction to the CPU than is using pinned memory for the tensor.  My bad.  I didn't know that the profiler would so distort the result.

However, there might still be value in allowing users to do a reduction into pinned memory.  CUB's reduce routines will certainly allow it.

So that's the whole thing.  I don't remember mentioning pageable anywhere...

What I'm trying to achieve is performance.

(We all work at the same company, right?  You can just send me and Matthias a note on slack if you have questions.)

**Comment by niuliling123** [comment]
I meet the same problem, during optimized ops(eg: AdamW), the op try to allocate host memory, but there's no pin_memory was free, it'll cost long time to wait
